auth/src/krb.cpp

```diff
@@ -0,0 +1,942 @@
+                    "reading token length: %d of %d bytes read", bytes_read, 4);
+            return ERROR(ec, fmt::format(fmt::runtime(
+                        "Error reading KRB token, length {} of {} bytes read."),
+                        bytes_read, 4));
+        }
+
+        length = ntohl( length );
+
+        if ( ikrbDebugFlag > 0 ) {
+            fprintf( stderr, "token length = %d\n", length );
+        }
+
+        *_rtn_length = length;
+
+        return SUCCESS();
+    } // krb_rcv_token_header
+
+    /// @brief Read a KRB token body
+    irods::error krb_rcv_token_body(int _fd,
+                                    gss_buffer_t _token,
+                                    unsigned int _length,
+                                    unsigned int* _rtn_bytes_read)
+    {
+        unsigned int bytes_read;
+
+        if (_token->length < _length) {
+            const auto ec = KRB_ERROR_TOKEN_TOO_LARGE;
+            const std::string msg = fmt::format(fmt::runtime(
+                    "[%s] error, token is too large for buffer, %d bytes in token, buffer is %d bytes"),
+                    __func__, _length, _token->length);
+            rodsLogAndErrorMsg(LOG_ERROR, ikrb_rErrorPtr, ec, msg.c_str());
+            return ERROR(ec, msg);
+        }
+
+        if (!_token->value) {
+            return ERROR(KRB_ERROR_BAD_TOKEN_RCVED, "Error KRB token buffer has NULL value.");
+        }
+
+        _token->length = _length;
+
+        if (const auto err = krb_read_all(_fd, (char*)_token->value, _token->length, &bytes_read); !err.ok()) {
+            return PASSMSG("Error reading KRB token body.", err);
+        }
+
+        if (bytes_read != _token->length) {
+            const auto ec = KRB_PARTIAL_TOKEN_READ;
+            const std::string msg = fmt::format(fmt::runtime(
+                    "Error reading token data, {} of {} bytes read."),
+                    bytes_read, _token->length);
+            rodsLogAndErrorMsg(LOG_ERROR, ikrb_rErrorPtr, ec, msg.c_str());
+            return ERROR(ec, msg);
+        }
+
+        *_rtn_bytes_read = _token->length;
+
+        return SUCCESS();
+    } // krb_rcv_token_body
+
+    ///@brief Receive a KRB token
+    irods::error krb_receive_token(int _fd, gss_buffer_t _token, unsigned int* _rtn_bytes_read)
+    {
+        int tmpLength;
+        char* cp;
+        int i;
+
+        if ( ikrbTokenHeaderMode ) {
+
+            /*
+              First, if in normal mode, peek to see if the other side is sending
+              headers and possibly switch into non-header mode.
+            */
+            tmpLength = 0;
+            cp = ( char * ) &tmpLength;
+            if ( sizeof( tmpLength ) > 4 ) {
+                cp += sizeof( tmpLength ) - 4;
+            }
+            i = recv( _fd, cp, 4, MSG_PEEK );
+            tmpLength = ntohl( tmpLength );
+            if ( ikrbDebugFlag > 0 ) {
+                fprintf( stderr, "peek length = %d\n", tmpLength );
```

Comment:
Any idea on what side of the connection this will fire?

We don't want calls to the printf-family of functions to run on the server side. Shouldn't hurt anything, but we won't see these messages in the log either.

Reply:
I believe this will mainly fire on the client side, but in a clerver situation (i.e. federation), it may fire on the server side. I can make an issue for this as this has always been here and continues to exist in the legacy plugin.

Reply:
I created https://github.com/irods/irods_auth_plugin_kerberos/issues/55 to deal with this.